client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -73,16 +74,19 @@
     private boolean closed;
     @GuardedBy("readers")
     private final List<EventSegmentReader> readers = new ArrayList<>();
+    // Ranges map is heavily used to build Position objects that are returned to the client. While there is no change
+    // in segment distribution, we reuse the same map instance for performance reasons. But this map should have a
+    // copy-on-write behavior to do not impact all the Position objects referencing the previous version of it.
     @GuardedBy("readers")
-    private final Map<Segment, Range> ranges = new HashMap<>();
+    private final AtomicReference<Map<Segment, Range>> ranges = new AtomicReference<>(new HashMap<>());
```

Comment:
This does not need to be an atomic reference, it is guarded by the readers lock, so it can be re-assigned.

Reply:
Done.